crates/astria-sequencer-relayer/src/relayer/write.rs

```diff
@@ -0,0 +1,366 @@
+        self.tx.send(block).await
+    }
+}
+
+pub(super) struct BlobSubmitter {
+    // The client to submit blobs to Celestia.
+    client: HttpClient,
+
+    // The channel over which sequencer blocks to receive.
+    blocks: Receiver<SequencerBlock>,
+
+    // The maximum number of conversions that can be active at the same time.
+    // Submitter will refuse to accept more blocks until there are fewer active
+    // conversions.
+    max_conversions: usize,
+
+    // The collection of tasks converting from sequencer blocks to celestia blobs,
+    // with the sequencer blocks' heights used as keys.
+    conversions: JoinMap<SequencerHeight, eyre::Result<Vec<Blob>>>,
+
+    // The maximum number of blobs permitted to sit in the blob queue.
+    // Submitter will refuse to accept more blocks until the queue is freed up again.
+    max_blobs: usize,
+
+    // Celestia blobs waiting to be submitted after conversion from sequencer blocks.
+    blobs: QueuedBlobs,
+
+    // The state of the relayer.
+    state: Arc<super::State>,
+}
+
+impl BlobSubmitter {
+    pub(super) fn new(client: HttpClient, state: Arc<super::State>) -> (Self, BlobSubmitterHandle) {
+        // XXX: The channel size here is just a number. It should probably be based on some
+        // heuristic about the number of expected blobs in a block.
+        let (tx, rx) = mpsc::channel(128);
+        let submitter = Self {
+            client,
+            blocks: rx,
+            max_conversions: 8,
+            conversions: JoinMap::new(),
+            max_blobs: 128,
+            blobs: QueuedBlobs::new(),
+            state,
+        };
+        let handle = BlobSubmitterHandle {
+            tx,
+        };
+        (submitter, handle)
+    }
+
+    pub(super) async fn run(mut self) -> eyre::Result<()> {
+        let mut submission = Fuse::terminated();
+
+        loop {
+            select!(
+                Some(block) = self.blocks.recv(), if self.has_capacity() => {
+                    debug!(
+                        height = %block.height(),
+                        "received sequencer block for submission",
+                    );
+                    self.conversions.spawn_blocking(block.height(), move || convert(block));
+                }
+
+                Some((sequencer_height, conversion_result)) = self.conversions.join_next() => {
+                     match crate::utils::flatten(conversion_result) {
+                        // XXX: Emitting at ERROR level because failing to convert contitutes
+                        // a fundamental problem for the relayer, even though it can happily
+                        // continue chugging along.
+                        // XXX: Should there instead be a mechanism to bubble up the error and
+                        // have sequencer-relayer return with an error code (so that k8s can halt
+                        // the chain)? This should probably be part of the protocal/sequencer proper.
```

Comment:
```suggestion
                        // the chain)? This should probably be part of the protocol/sequencer proper.
```

yeah, failing to convert would be very bad - in what cases would that ever happen?